src/evaluators/BinaryExpression.js

```diff
@@ -65,6 +65,7 @@ export function getPureBinaryOperationResultType(
     if (realm.isInPureScope()) {
       if (!leftPure) Leak.leakValue(realm, lval);
       if (!rightPure) Leak.leakValue(realm, rval);
+      // These are not guaranteed to be pure, is this type valid? If it is even when not pure then it is misnamed.
```

Comment:
Looking through the current callsites, it seems to be that `typeIfPure` is always correct even if the arguments are non-pure.